os/pm/pm_timer/pm_timer_activity.c

```diff
@@ -0,0 +1,244 @@
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name NuttX nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ************************************************************************/
+
+/************************************************************************
+ * Included Files
+ ************************************************************************/
+
+#include <tinyara/pm/pm.h>
+#include <tinyara/irq.h>
+
+#include "pm_timer.h"
+
+#define MIN(a, b) (((a) < (b)) ? (a) : (b))
+
+/************************************************************************
+ * Pre-processor Definitions
+ ************************************************************************/
+
+/************************************************************************
+ * Private Type Declarations
+ ************************************************************************/
+
+/************************************************************************
+ * Public Variables
+ ************************************************************************/
+
+/************************************************************************
+ * Private Functions
+ ************************************************************************/
+
+void pm_timer_expire(void) 
+{
+	pm_wakeup_timer_t *head;
+	irqstate_t state;
+
+	while (g_pmTimer_activeList.head && ((pm_wakeup_timer_t *)g_pmTimer_activeList.head)->delay <= 0) {
+		
+    		head = (pm_wakeup_timer_t *)g_pmTimer_activeList.head;
+
+		/* Check if this needs to be atomic */
+		state = enter_critical_section();
+
+		/* Remove the pm timer from active list and lock the system 
+		 * for the specific pid, so that the specific thread can complete 
+		 * its function */
+		sq_remfirst(&g_pmTimer_activeList);
+		if (is_pm_lock[head->pid] == PM_PID_UNLOCK) {
+			is_pm_lock[head->pid] = PM_PID_LOCK;
+			pm_stay(PM_IDLE_DOMAIN, PM_NORMAL);
+			pmdbg("pm is locked for process with id = %d\n", head->pid);
+		}
+
+		if (PM_ISSTATIC(head)) {
```

Comment:
If we have create fucntion, we need to remove or deleate fucntion also. It is pair.

Please make a pm_timer_delete function and call here